src/ucp/core/ucp_ep.c

```diff
@@ -139,36 +149,67 @@ ucs_status_t ucp_ep_create_base(ucp_worker_h worker, const char *peer_name,
     ucs_debug("created ep %p to %s %s", ep, ucp_ep_peer_name(ep), message);
     return UCS_OK;
 
+err_free_keys:
+    ucs_free(ucp_ep_ext_gen(ep)->ids);
 err_free_ep:
     ucs_strided_alloc_put(&worker->ep_alloc, ep);
 err:
     return status;
 }
 
-ucs_status_t ucp_worker_create_ep(ucp_worker_h worker, const char *peer_name,
-                                  const char *message, ucp_ep_h *ep_p)
+static void ucp_ep_destroy_base(ucp_ep_h ep)
+{
+    ucs_free(ucp_ep_ext_gen(ep)->ids);
+    ucs_strided_alloc_put(&ep->worker->ep_alloc, ep);
+}
+
+ucs_status_t ucp_worker_create_ep(ucp_worker_h worker, unsigned ep_init_flags,
+                                  const char *peer_name, const char *message,
+                                  ucp_ep_h *ep_p)
 {
     ucs_status_t status;
     ucp_ep_h ep;
 
     status = ucp_ep_create_base(worker, peer_name, message, &ep);
     if (status != UCS_OK) {
-        return status;
+        goto err;
     }
 
     ucs_list_add_tail(&worker->all_eps, &ucp_ep_ext_gen(ep)->ep_list);
 
+    ucs_assert(ucp_ep_ext_gen(ep)->ids->local == UCP_EP_ID_INVALID);
+    status = ucs_ptr_map_put(&worker->ptr_map, ep,
+                             !!(ep_init_flags &
+                                UCP_EP_INIT_ERR_MODE_PEER_FAILURE),
+                             &ucp_ep_ext_gen(ep)->ids->local);
+    if (status != UCS_OK) {
+        goto err_destroy_ep_base;
+    }
+
     *ep_p = ep;
 
     return UCS_OK;
+
+err_destroy_ep_base:
+    ucp_ep_destroy_base(ep);
+err:
+    return status;
 }
 
 void ucp_ep_delete(ucp_ep_h ep)
 {
+    ucs_status_t status UCS_V_UNUSED;
+
     ucs_callbackq_remove_if(&ep->worker->uct->progress_q,
                             ucp_wireup_msg_ack_cb_pred, ep);
     UCS_STATS_NODE_FREE(ep->stats);
     ucs_list_del(&ucp_ep_ext_gen(ep)->ep_list);
+    if (ucp_ep_ext_gen(ep)->ids->local != UCP_EP_ID_INVALID) {
```

Comment:
why would it be invalid?

Reply:
if it's tmp_ep

Reply:
1. another reason that tmp_ep is bad, need to add extra checks in unrelated places
2. which function created the tmp_ep? need to destroy tmp_ep with the opposite of that function